File: sdks/python/src/osva/resources/approvals.py
from __future__ import annotations

from typing import Any

from osva.http_client import HTTPClient


class ApprovalsResource:
    def __init__(self, client: HTTPClient) -> None:
        self._client = client

    def get(self, approval_request_id: str) -> dict[str, Any]:
        return self._client.request(
            "GET",
            f"/v1/approval-requests/{approval_request_id}",
        )

    def decide(self, approval_request_id: str, payload: dict[str, Any]) -> dict[str, Any]:
        return self._client.request(
            "POST",
            f"/v1/approval-requests/{approval_request_id}/decision",
            body=payload,
        )

File: packages/sdk/src/resources/approvals.ts
import type { ApprovalRequestId } from "@osva/contracts";
import {
  approvalRequestResourceSchema,
  decideApprovalRequestSchema,
} from "@osva/contracts/schemas";
import type { z } from "zod";

import type { OsvaHttpClient } from "../http-client.js";

type ApprovalRequestResource = z.infer<typeof approvalRequestResourceSchema>;
type DecideApprovalRequest = z.infer<typeof decideApprovalRequestSchema>;

export class ApprovalsResource {
  constructor(private readonly client: OsvaHttpClient) {}

  get(approvalRequestId: ApprovalRequestId): Promise<ApprovalRequestResource> {
    return this.client.request({
      method: "GET",
      path: `/v1/approval-requests/${encodeURIComponent(approvalRequestId)}`,
    });
  }

  decide(
    approvalRequestId: ApprovalRequestId,
    input: DecideApprovalRequest,
  ): Promise<ApprovalRequestResource> {
    return this.client.request({
      method: "POST",
      path: `/v1/approval-requests/${encodeURIComponent(approvalRequestId)}/decision`,
      body: input,
    });
  }
}
